File: services/lpo-service/audit_middleware.py
import datetime
import json
import os
import re
import threading
import urllib.request

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request

# --- OpenTelemetry (SPEC w9 addendum): audit_ship_failures_total counter ---
import sys as _otel_sys

_otel_sys.path.insert(
    0,
    os.path.normpath(
        os.path.join(os.path.dirname(__file__), "..", "..", "shared", "otel", "python")
    ),
)
try:
    from otelkit import inc_counter
except Exception:

    def inc_counter(name, attrs=None):
        return None

_AUDIT_URL = os.getenv("AUDIT_SVC_URL", "http://audit-service:8000")
# AU-01 (F15-1): shared-secret ingest credential (k8s secret audit-ingest-credentials).
_AUDIT_INGEST_TOKEN = os.getenv("AUDIT_INGEST_TOKEN", "")
_SKIP_METHODS = {"GET", "HEAD", "OPTIONS"}
_SKIP_PREFIXES = ("/health", "/metrics", "/dapr", "/docs", "/openapi")
_UUID_RE = re.compile(r"/[0-9a-f]{8}(?:-[0-9a-f]{4}){3}-[0-9a-f]{12}")
_INT_RE = re.compile(r"/\d+")


def _path_to_event_type(method: str, path: str) -> str:
    clean = _UUID_RE.sub("/{id}", path)
    clean = _INT_RE.sub("/{id}", clean)
    return f"{method}:{clean}"


def _emit(actor_id: str, tenant_id: str, event_type: str, event_data: dict) -> None:
    if not _AUDIT_URL:
        return
    try:
        body = json.dumps({
            "actor_id": actor_id,
            "tenant_id": tenant_id,
            "event_type": event_type,
            "event_data": event_data,
            "timestamp": datetime.datetime.utcnow().isoformat(),
        }).encode()
        req = urllib.request.Request(
            f"{_AUDIT_URL}/audits",
            data=body,
            headers={
                "Content-Type": "application/json",
                "x-tenant-id": tenant_id,
                "x-keycloak-id": "system",
            },
            method="POST",
        )
        # AU-01 (F15-1): present the shared ingest credential so the fail-closed
        # audit-service accepts the event (secret: audit-ingest-credentials).
        if _AUDIT_INGEST_TOKEN:
            req.add_header("X-Audit-Ingest-Token", _AUDIT_INGEST_TOKEN)
        urllib.request.urlopen(req, timeout=3)
    except Exception:
        # Alerting contract (w9 addendum): count swallowed audit ship failures.
        inc_counter(
            "audit_ship_failures_total",
            {"service": "lpo-service", "tenant_id": tenant_id or "unknown"},
        )


class AuditMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        if request.method in _SKIP_METHODS:
            return await call_next(request)
        path = request.url.path
        if any(path.startswith(p) for p in _SKIP_PREFIXES):
            return await call_next(request)

        response = await call_next(request)

        tenant_id = request.headers.get("x-tenant-id", "unknown")
        actor_id = (
            request.headers.get("x-keycloak-id")
            or request.headers.get("x-keycloak-id", "unknown")
        )
        event_data: dict = {
            "method": request.method,
            "path": path,
            "status_code": response.status_code,
        }
        if str(request.query_params):
            event_data["query"] = str(request.query_params)

        threading.Thread(
            target=_emit,
            args=(actor_id, tenant_id, _path_to_event_type(request.method, path), event_data),
            daemon=True,
        ).start()

        return response

File: services/chart-of-accounts-service/audit_middleware.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

var (
	auditSvcURL      = os.Getenv("AUDIT_SVC_URL")
	auditIngestToken = os.Getenv("AUDIT_INGEST_TOKEN") // AU-01
	skipPrefixes     = []string{"/health", "/metrics", "/dapr", "/docs", "/ready"}
	skipMethods      = map[string]bool{"GET": true, "HEAD": true, "OPTIONS": true}
	auditUUIDRE      = regexp.MustCompile(`/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`)
	auditIntRE       = regexp.MustCompile(`/[0-9]+`)
)

// auditShipFailures implements the w9 alerting contract counter
// audit_ship_failures_total{service,tenant_id} for audit shipping failures.
var auditShipFailures = promauto.NewCounterVec(
	prometheus.CounterOpts{
		Name: "audit_ship_failures_total",
		Help: "Audit event shipping failures (w9 alerting contract).",
	},
	[]string{"service", "tenant_id"},
)

func init() {
	if auditSvcURL == "" {
		auditSvcURL = "http://audit-service:8000"
	}
}

type auditResponseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (w *auditResponseWriter) WriteHeader(code int) {
	w.statusCode = code
	w.ResponseWriter.WriteHeader(code)
}

func auditPathToEventType(method, path string) string {
	clean := auditUUIDRE.ReplaceAllString(path, "/{id}")
	clean = auditIntRE.ReplaceAllString(clean, "/{id}")
	return fmt.Sprintf("%s:%s", method, clean)
}

func sendAuditEvent(actorID, tenantID, eventType string, eventData map[string]interface{}) {
	payload, err := json.Marshal(map[string]interface{}{
		"actor_id":   actorID,
		"tenant_id":  tenantID,
		"event_type": eventType,
		"event_data": eventData,
		"timestamp":  time.Now().UTC().Format(time.RFC3339),
	})
	if err != nil {
		return
	}
	req, err := http.NewRequest("POST", auditSvcURL+"/audits", bytes.NewBuffer(payload))
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-tenant-id", tenantID)
	req.Header.Set("x-keycloak-id", "system")
	if auditIngestToken != "" {
		// AU-01 (F15-1): shared ingest credential; audit-service fails closed without it.
		req.Header.Set("X-Audit-Ingest-Token", auditIngestToken)
	}
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		auditShipFailures.WithLabelValues("chart-of-accounts-service", tenantID).Inc()
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		auditShipFailures.WithLabelValues("chart-of-accounts-service", tenantID).Inc()
	}
}

func auditMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if skipMethods[r.Method] {
			next.ServeHTTP(w, r)
			return
		}
		for _, p := range skipPrefixes {
			if strings.HasPrefix(r.URL.Path, p) {
				next.ServeHTTP(w, r)
				return
			}
		}

		rec := &auditResponseWriter{ResponseWriter: w, statusCode: http.StatusOK}
		next.ServeHTTP(rec, r)

		actorID := r.Header.Get("X-Keycloak-ID")
		if actorID == "" {
			actorID = r.Header.Get("x-keycloak-id")
		}
		if actorID == "" {
			actorID = "unknown"
		}
		tenantID := r.Header.Get("X-Tenant-ID")
		if tenantID == "" {
			tenantID = "unknown"
		}
		eventData := map[string]interface{}{
			"method":      r.Method,
			"path":        r.URL.Path,
			"status_code": rec.statusCode,
		}
		if r.URL.RawQuery != "" {
			eventData["query"] = r.URL.RawQuery
		}

		go sendAuditEvent(actorID, tenantID, auditPathToEventType(r.Method, r.URL.Path), eventData)
	})
}
